x/subscription/keeper/grpc_query_estimated_rewards.go

```diff
@@ -0,0 +1,128 @@
+	var entry epochstoragetypes.StakeEntry
+	found := false
+	for _, e := range storage {
+		totalStake = totalStake.Add(e.EffectiveStake())
+		if e.Address == req.Provider {
+			found = true
+			entry = e
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("provider not found in stake entries for chain ID: %s", req.ChainId)
+	}
+
+	// delegation of the provider
+	delegationAmount := entry.Stake
+	delegatorPart := sdk.NewDecFromInt(delegationAmount.Amount).QuoInt(totalStake).MulInt64(int64(entry.DelegateCommission)).QuoInt64(100)
+	if req.Delegator != "" || req.Amount != "" {
+		if req.Amount != "" {
+			var err error
+			delegationAmount, err = sdk.ParseCoinNormalized(req.Amount)
+			if err != nil {
+				return nil, err
+			}
+			entry.DelegateTotal = entry.DelegateTotal.Add(delegationAmount)
+		} else if req.Delegator != "" {
+			d, found := k.dualstakingKeeper.GetDelegation(ctx, req.Delegator, req.Provider, req.ChainId, uint64(ctx.BlockHeight()))
+			if !found {
+				return nil, fmt.Errorf("could not find delegator")
+			}
+			delegationAmount = d.Amount
+		}
+
+		delegatorPart = sdk.NewDecFromInt(delegationAmount.Amount).QuoInt(totalStake).MulInt64(int64(100 - entry.DelegateCommission)).QuoInt64(100)
+		// calculate existing delegator rewards
+		if entry.DelegateLimit.Amount.LT(entry.DelegateTotal.Amount) {
+			delegatorPart = delegatorPart.MulInt(entry.DelegateLimit.Amount).QuoInt(entry.DelegateTotal.Amount)
+		}
+	}
+	delegatorPart = delegatorPart.Mul(sdk.OneDec().Sub(specContribut))
+
+	totalSubsRewards := sdk.Coins{}
+	subsIndices := k.GetAllSubscriptionsIndices(ctx)
+	for _, subIndex := range subsIndices {
+		sub, found := k.GetSubscription(ctx, subIndex)
+		if found {
+			sub.Credit.Amount = sub.Credit.Amount.Quo(sdk.NewIntFromUint64(sub.DurationLeft))
+			totalSubsRewards = totalSubsRewards.Add(sub.Credit)
+		}
+	}
+
+	emissions := k.rewardsKeeper.SpecEmissionParts(ctx)
+	var specEmission rewardstypes.SpecEmissionPart
+	found = false
+	for _, k := range emissions {
+		if k.ChainID == req.ChainId {
+			specEmission = k
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("spec emission part not found for chain ID: %s", req.ChainId)
+	}
+
+	subscriptionRewards := sdk.NewCoins(totalSubsRewards.MulInt(specEmission.Emission.MulInt64(100).RoundInt())...)
+	valRewards, comRewards, err := k.rewardsKeeper.CalculateValidatorsAndCommunityParticipationRewards(ctx, subscriptionRewards[0])
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate Validators And Community Participation Rewards")
+	}
+
+	subscriptionRewards = subscriptionRewards.Sub(valRewards...).Sub(comRewards...)
+
+	coins := k.rewardsKeeper.TotalPoolTokens(ctx, rewardstypes.ProviderRewardsDistributionPool)
+	TotalPoolTokens := coins.AmountOf(k.stakingKeeper.BondDenom(ctx))
+
+	bonusRewards := k.rewardsKeeper.SpecTotalPayout(ctx,
+		TotalPoolTokens,
+		sdk.NewDecFromInt(subscriptionRewards.AmountOf(k.stakingKeeper.BondDenom(ctx))),
```

Comment:
the calculation of the boost is before the validator tax, so do this part before this line: subscriptionRewards = subscriptionRewards.Sub(valRewards...).Sub(comRewards...)

Reply:
done